### DISCUSSION TITLE
How can I make a dataset?

### DISCUSSION THREAD

**Discussion by kdhjk1** [moshi4/pyGenomeViz#11] [Q&A] [answered]
Loading dataset is the first start of pygenomeviz of API.
How can I make dataset with gbk files and links? 
And please write the practical loading procedures for jupyter notebook.

**Comment by moshi4** [comment, marked as answer]
The procedure for visualizing multiple Genbank files and MUMmer comparison link results is described as jupyter notebooks in the following document.
https://moshi4.github.io/pyGenomeViz/getting_started/#2-4-genome-alignment-visualization-workflow

The source code of [pgv-mummer](https://github.com/moshi4/pyGenomeViz/blob/main/src/pygenomeviz/scripts/mummer.py) and [pgv-mmseqs](https://github.com/moshi4/pyGenomeViz/blob/main/src/pygenomeviz/scripts/mmseqs.py) CLI may also help you understand the dataset preparation procedure.

**Comment by kdhjk1** [comment]


**Reply by moshi4** [reply]
> I tried some scripts and have a question now for the pygenomeviz.
> Can I draw the figure for the selected range like the Gallery Example. [3. Enterobacteria phage](https://moshi4.github.io/pyGenomeViz/api-examples/gallery_example/#3-enterobacteria-phage)

Genome selected regions can be plotted by setting `min_range` and `max_range` in the `Genbank` parser class and the option `start_pos`(=gbk.min_range) in the `add_feature_track` function.

Below is an example code that plots only NC_019724 and NC_013600 selected regions among multiple genomes.

```python
from pygenomeviz import Genbank, GenomeViz, load_dataset

gv = GenomeViz(
    fig_width=10,
    fig_track_height=0.5,
    feature_track_ratio=0.5,
    tick_track_ratio=0.5,
    align_type="center",
    tick_style="bar",
    tick_labelsize=10,
)

gbk_files, links = load_dataset("enterobacteria_phage")
for idx, gbk_file in enumerate(gbk_files):
    if idx in (0, 3):  
        # Limit NC_019724, NC_013600 plot genomic range (10,000 - 30,000 bp)
        gbk = Genbank(gbk_file, min_range=10000, max_range=30000)
    else:
        gbk = Genbank(gbk_file)

    track = gv.add_feature_track(gbk.name, size=gbk.range_size, start_pos=gbk.min_range, labelsize=10)
    track.add_genbank_features(
        gbk,
        label_type="product" if idx == 0 else None,  # Labeling only top track
        label_handle_func=lambda s : "" if s.startswith("hypothetical") else s,  # Ignore 'hypothetical ~~~' label
        labelsize=8,
        labelvpos="top",
        facecolor="skyblue",
        linewidth=0.5,
        labelrotation=45,
    )
    track.set_sublabel(size=8)

fig = gv.plotfig()
```

![output](https://user-images.githubusercontent.com/41852815/211299857-955d3487-9233-4139-bd13-a81eeb49c561.png)

> Whole genome drawing is working but it is not working with some error messages such as self.start & self.end.

If you provide me the code in which the error occurred, I will be able to give you more specific advice.

**Comment by kdhjk1** [comment]


**Reply by moshi4** [reply]
Hi, Jae-Seok Kim

I am a Japanese who majored in molecular evolution and bioinformatics as a graduate student several years ago, and currently work for a software development company unrelated to bioinformatics.

I am developing several tools including pyGenomeViz on GitHub with the main purpose of improving my OSS and software development skills. 
Since the bioinformatics field was easy for me personally to handle as a subject for software development, I am mainly trying to develop comparative genomics visualization tools at the moment.

As mentioned above, I am not currently in a career as a researcher, but as a software developer.
Therefore, I am sorry, but it is difficult for me to participate in the research discussion you have proposed.

Of course, if you have any questions about the software I have developed, I will respond to the extent that I can answer them, so please feel free to ask.

Best Regards,

**Comment by kdhjk1** [comment]


**Reply by moshi4** [reply]
Hi, 

> I attached a document file and an ipynb file.

I assume you are replying to an email from GitHub.
Actually, in this case, the file attached to the email is ignored, so the attachment is not visible to me.

Is it possible to upload the file directly to this Discussions on GitHub?
https://github.com/moshi4/pyGenomeViz/discussions/11

**Comment by kdhjk1** [comment]
The drawing with whole genome is possible.  
Error message when drawing whole genome with range option (ex. min_range=10000, max_range=25000)

Can I draw some link lines with selected ranges with genome sequence, not the whole range?

[pygenomeviz error question.odt](https://github.com/moshi4/pyGenomeViz/files/10534754/pygenomeviz.error.question.odt)
[figure .pdf](https://github.com/moshi4/pyGenomeViz/files/10534763/figure.pdf)

**Reply by moshi4** [reply]
Since you have not presented the source code, I do not know the cause of the error.
However, it is possible to compare and visualize the specified range genomes with the CLI of [`pgv-mummer`](https://moshi4.github.io/pyGenomeViz/cli-docs/pgv-mummer/#example-2) and [`pgv-mmseqs`](https://moshi4.github.io/pyGenomeViz/cli-docs/pgv-mmseqs/#example-2), but did you try it?

Here is an example of CLI.

```shell
# Download test dataset
pgv-download-dataset -n enterobacteria_phage
# Compare & visualize 3 species genome with range in 10000 - 25000 bp
pgv-mummer --gbk_resources NC_019724.gbk:10000-25000 NC_024783.gbk:10000-25000 NC_016566.gbk:10000-25000 -o mummer_example
```

![result](https://user-images.githubusercontent.com/41852815/215456754-2523d735-e0c1-4691-aac7-be15edc4324f.png)

If you want a more custom pgv-mummer output figure from the API, you can modify the source in [src/pygenomeviz/mummer.py](https://github.com/moshi4/pyGenomeViz/blob/main/src/pygenomeviz/scripts/mummer.py). The `pgv-mummer` command is synonymous with `python mummer.py`.

**Comment by kdhjk1** [comment]
Thank you very much! I will try it. 

And, I think it would be so good to draw the figure in page 2 of the attached pygenomeviz error question.odt file?
The whole genome can be drawn with some complicated labels, but it may cause error when I tried it with range limits (page 3).

**Reply by moshi4** [reply]
Oh, sorry, I overlooked the fact that the ODT file has multiple pages.

It seems to me that the error is due to an incorrectly specified argument in the `gv.add_feature_track` method on **line 26**.

Try changing the source code as follows.

```python
# Before Change (line 26)
track = gv.add_feature_track(gbk.name, gbk.range_size, labelsize=10)
```

```python
# After Change (line 26)
track = gv.add_feature_track(gbk.name, size=gbk.range_size, start_pos=gbk.min_range, labelsize=10)
```

This would resolve the error that is currently being displayed.

___

I think that even after resolving the above error, there is still a possibility that the `gv.add_link` method may cause an error.
If the error occurs in the `gv.add_link` method, please try the following source code change.

```python
# Before Change (line 43)
gv.add_link(link_data1, link_data2, ...)
```

```python
# After Change (line 43)
try:
    gv.add_link(link_data1, link_data2, ...)
except ValueError:
    # Ignore links outside the illegal range
    pass
```
